[sycl-post-link] Call cleanup at the very end of sycl-post-link

asudarsa opened the pull request:
This PR resolves https://github.com/intel/llvm/issues/9962

This change calls cleanup as the final step in sycl-post-link before writing the module into file. This helps to remove several dead variables. Some of these 'dead' variables seem to get translated to erroneous SPIR-V code resulting in spirv-val failures.
In addition to improving code size, this change help to resolve such errors as a side-effect.

Thanks
Sincerely

asudarsa commented:
@sarnex 

Can you please take a look? I tried removing the cleanup calls inside the ModuleSpiltter.cpp code, but that resulted in one of the ESIMD LIT tests failing. So I have kept them as is.

Thanks

asudarsa commented:
Hi @LU-JOHN

Can you please take a look?

Thanks

LU-JOHN commented on llvm/tools/sycl-post-link/ModuleSplitter.cpp:
```diff
@@ -314,6 +314,10 @@ ModuleDesc extractCallGraph(const ModuleDesc &MD,
 
   ModuleDesc SplitM = extractSubModule(MD, GVs, std::move(ModuleEntryPoints));
+  // TODO: cleanup pass is now called for each output module at the end of
+  // sycl-post-link. This call is redundant. Hiwever, we subsequently run
```
"Hiwever" is misspelled.

LU-JOHN commented:
> This helps to remove several dead variables.

Can we add a testcase that relies on this change to remove dead variables?

LU-JOHN commented:
> Some of these 'dead' variables seem to get translated to erroneous SPIR-V code resulting in spirv-val failures.
> In addition to improving code size, this change help to resolve such errors as a side-effect.

Do we know why a dead variable is translated to erroneous SPIR-V code?   I think this should be fixed so that the code works even without the cleanup in this change.

asudarsa commented:
> > Some of these 'dead' variables seem to get translated to erroneous SPIR-V code resulting in spirv-val failures.
> > In addition to improving code size, this change help to resolve such errors as a side-effect.
> 
> Do we know why a dead variable is translated to erroneous SPIR-V code? I think this should be fixed so that the code works even without the cleanup in this change.

Erroneous SPIR-V Code is generated as we are not generating these global variables with the correct address space. In trying to fix this issue, it was found that such variables can be actually removed at the end of sycl-post-link as they have 'private' linkage and are not used in the module where they are defined.
However, I do agree SPIR-V Translator needs to react the right way instead of generating wrong code. That will be tracked elsewhere.
This PR has it own benefits as we remove some dead code and this helps to avoid code bloating a bit.

Thanks

asudarsa commented:
> > This helps to remove several dead variables.
> 
> Can we add a testcase that relies on this change to remove dead variables?

Added a test. Infact, adding the test exposed a bug in the implementation. Fixed it. Thanks @LU-JOHN

LU-JOHN commented:
> > > This helps to remove several dead variables.
> > 
> > 
> > Can we add a testcase that relies on this change to remove dead variables?
> 
> Added a test. Infact, adding the test exposed a bug in the implementation. Fixed it. Thanks @LU-JOHN

Will this change also delete unused private functions?  If so, can the testcase show this?

asudarsa commented:
> > > > This helps to remove several dead variables.
> > > 
> > > 
> > > Can we add a testcase that relies on this change to remove dead variables?
> > 
> > 
> > Added a test. Infact, adding the test exposed a bug in the implementation. Fixed it. Thanks @LU-JOHN
> 
> Will this change also delete unused private functions? If so, can the testcase show this?

I do not suppose so. Main focus for this change is to remove Spec Consts once they have been consumed by sycl-post-link.

Thanks

AlexeySachkov commented:
HIP pre-commit failure is reported in #11405

I don't think that this change should cause any HIP-specific issues, so I will proceed with merge
